Static file generation makes app startup indeterminate

While looking at the logs of a `portal-backend` pod in the midst of spinning up, I see the following stack trace:
```
[2024-12-06 17:36:39 +0000] [40] [ERROR] Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/starlette/routing.py", line 693, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "/usr/local/lib/python3.9/contextlib.py", line 181, in __aenter__
    return await self.gen.__anext__()
  File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 133, in merged_lifespan
    async with original_context(app) as maybe_original_state:
  File "/usr/local/lib/python3.9/contextlib.py", line 181, in __aenter__
    return await self.gen.__anext__()
  File "/usr/local/lib/python3.9/site-packages/fastapi/routing.py", line 133, in merged_lifespan
    async with original_context(app) as maybe_original_state:
  File "/usr/local/lib/python3.9/contextlib.py", line 181, in __aenter__
2024-12-06T17:36:39.975833334Z     return await self.gen.__anext__()
  File "/app/nmdc_server/app.py", line 45, in lifespan
    generate_and_mount_static_files()
  File "/app/nmdc_server/app.py", line 40, in generate_and_mount_static_files
2024-12-06T17:36:39.975869755Z     generate_submission_schema_files(directory=static_path)
  File "/app/nmdc_server/static_files.py", line 43, in generate_submission_schema_files
    shutil.copyfile(str(gold_tree_path), out_dir / "GoldEcosystemTree.json")
  File "/usr/local/lib/python3.9/shutil.py", line 266, in copyfile
    with open(dst, 'wb') as fdst:
FileNotFoundError: [Errno 2] No such file or directory: 'static/submission_schema/GoldEcosystemTree.json'
```

This is also getting caught by sentry.

I believe that what's happening is several workers are all trying to generate and mount the static files into a directory, but one of the steps taken to do so is to delete the existing directory. This would lead to a race condition between the workers:

1. Worker A starts by removing the existing static files directory
2. Worker A creates a new static files directory
3. Worker B starts by removing the existing static files directory
4. Worker A tries to add a file to the not-yet-created static files directory

The functions in question are located in [nmdc_server/static_files.py](https://github.com/microbiomedata/nmdc-server/blob/main/nmdc_server/static_files.py), and get called in [nmdc_server/app.py::create_app](https://github.com/microbiomedata/nmdc-server/blob/5c1ce8fc59f8f0705c861cf40c8152e417148529/nmdc_server/app.py#L37-L40).

One quick fix might be to enable creating parent directories when creating the submission schema directory, i.e.

```
out_dir = directory / "submission_schema"
out_dir.mkdir(exist_ok=True
```

becomes

```
out_dir = directory / "submission_schema"
out_dir.mkdir(parents=True, exist_ok=True)
```

A better approach might be to move the creating of the static directory and static submission schema files into a CLI subcommand, and call that in `prestart.sh`, this way the static mounting only happens once.

--- a/nmdc_server/app.py
+++ b/nmdc_server/app.py
@@ -1,6 +1,5 @@
 import logging
 import typing
-from contextlib import asynccontextmanager
 
 import sentry_sdk
 from debug_toolbar.middleware import DebugToolbarMiddleware
@@ -15,7 +14,7 @@ from starlette.middleware.sessions import SessionMiddleware
 from nmdc_server import __version__, api, auth, errors
 from nmdc_server.config import settings
 from nmdc_server.database import after_cursor_execute, before_cursor_execute, listen
-from nmdc_server.static_files import generate_submission_schema_files, initialize_static_directory
+from nmdc_server.static_files import static_path
 
 
 def attach_sentry(app: FastAPI):
@@ -42,16 +41,6 @@ def create_app(env: typing.Mapping[str, str]) -> FastAPI:
         listen(Engine, "before_cursor_execute", before_cursor_execute)
         listen(Engine, "after_cursor_execute", after_cursor_execute)
 
-    def generate_and_mount_static_files():
-        static_path = initialize_static_directory(remove_existing=True)
-        generate_submission_schema_files(directory=static_path)
-        app.mount("/static", StaticFiles(directory=static_path), name="static")
-
-    @asynccontextmanager
-    async def lifespan(app: FastAPI):
-        generate_and_mount_static_files()
-        yield
-
     app = FastAPI(
         title="NMDC Data and Submission Portal API",
         description="""
@@ -64,8 +53,10 @@ field and click "Authorize".
         docs_url="/api/docs",
         openapi_url="/api/openapi.json",
         debug=settings.debug,
-        lifespan=lifespan,
     )
+    if not static_path.is_dir():
+        raise Exception("Static files not found")
+    app.mount("/static", StaticFiles(directory=static_path), name="static")
     if settings.environment == "development":
         app.add_middleware(
             DebugToolbarMiddleware,

--- a/nmdc_server/cli.py
+++ b/nmdc_server/cli.py
@@ -14,6 +14,7 @@ from nmdc_server import jobs
 from nmdc_server.config import Settings
 from nmdc_server.database import SessionLocalIngest
 from nmdc_server.ingest import errors
+from nmdc_server.static_files import generate_submission_schema_files, initialize_static_directory
 
 
 def send_slack_message(text: str) -> bool:
@@ -339,6 +340,16 @@ def load_db(key_file, user, host, list_backups, backup_file):
         sys.exit(1)
 
     click.secho(f"\nSuccessfully loaded {settings.current_db_uri}", fg="green")
+
+
+@cli.command()
+@click.option("--remove-existing", is_flag=True, default=False)
+def generate_static_files(remove_existing):
+    click.echo("Generating static files...")
+    static_path = initialize_static_directory(remove_existing=remove_existing)
+    click.echo("Generating submission schema files...")
+    generate_submission_schema_files(directory=static_path)
+    click.echo("Done generating static files.")
 
 
 if __name__ == "__main__":

--- a/nmdc_server/static_files.py
+++ b/nmdc_server/static_files.py
@@ -5,12 +5,20 @@ from pathlib import Path
 from linkml_runtime import SchemaView
 from linkml_runtime.dumpers import json_dumper
 
+static_path = Path("static")
+
 
 def initialize_static_directory(*, remove_existing=False) -> Path:
-    static_path = Path("static")
     if remove_existing:
+        # Delete existing contents of the static directory,
+        # but keep the directory itself, since it may already
+        # be mounted by the FastAPI app.
         try:
-            shutil.rmtree(static_path)
+            for child in static_path.iterdir():
+                if child.is_dir():
+                    shutil.rmtree(child)
+                else:
+                    child.unlink()
         except FileNotFoundError:
             pass
     static_path.mkdir(parents=True, exist_ok=True)

--- a/start.sh
+++ b/start.sh
@@ -18,5 +18,8 @@ PGDATABASE=postgres psql -c "create database nmdc_b;" || true
 # Apply pending alembic migrations
 nmdc-server migrate
 
+# Generate static content
+nmdc-server generate-static-files --remove-existing
+
 ## Start the server
 uvicorn nmdc_server.asgi:app --host 0.0.0.0 --port 8000

--- a/tox.ini
+++ b/tox.ini
@@ -17,7 +17,9 @@ setenv =
   SQLALCHEMY_SILENCE_UBER_WARNING=1
 passenv =
   NMDC_TESTING_DATABASE_URI
-commands = pytest {posargs} tests/
+commands =
+    nmdc-server generate-static-files --remove-existing
+    pytest {posargs} tests/
 
 [testenv:ingest]
 setenv =

